Add generic claim extractor to get claims from ID Tokens

JoelSpeed opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description

This PR replaces all ID Token claim extraction with a single claim extractor utility.
The utility is designed to first search the ID Token for a named claim, and, if not present, fetch the claim from the profile URL if configured.

## Motivation and Context

This should make it easier to integrate additional claims in the future. A long requested feature is to have a generic map of "extra" claims, and with this change, this should now be achievable.

## How Has This Been Tested?

<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Unit testing of the utility itself. Some manual testing using the local testing environment.

Marked as WIP as I'd like to decide what to do about some of the tests that I currently have commented out.

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] My change requires a change to the documentation or CHANGELOG.
- [x] I have updated the documentation/CHANGELOG accordingly.
- [x] I have created a feature (non-master) branch for my PR.

kragniz reviewed (approved):
There's a fair amount of movement here, but overall looks good

JoelSpeed commented:
Thanks for the review @kragniz, I've rebased which applied cleanly and added a changelog entry, otherwise there were no changes, will merge based on your prior approval
